[v1.15] route: Specify "proto kernel" for ip routes and rules

jschwinger233 opened the pull request:
* [x] #31716 @jschwinger233 

This is a customized backport to take care of upgrade/downgrade. Please see commit message for details.

jschwinger233 committed:
route: Specify "proto kernel" for ip routes and rules

[ upstream commit: 318a64874832bbc9c0047c114a0723d8bfb9219d ]

v1.14 installs routes and rules with specific "proto kernel", v1.15
missed them. Without "proto kernel", it causes troubles when downgrade
from v1.15 to v1.14, as v1.14 is deleting routes with "proto kernel" but
there are no matching ones.

Signed-off-by: Zhichuan Liang <gray.liang@isovalent.com>
Signed-off-by: Zhichuan Liang <gray.liang@isovalent.com>

julianwiedmann commented on pkg/proxy/routes.go:
```diff
@@ -196,3 +207,26 @@ func removeFromProxyRoutesIPv6() error {
+		}
+	}
+	return nil
+}
```
Would it help for review to split the custom cleanup logic into a separate patch (and make the first patch a clean cherry-pick from upstream) ? Or is it crucial for bisectability that all these changes go in as one patch?

jschwinger233 replied:
Seems no harm to split into two patches. 1.15-old -> 1.15-tip upgrade is the only one that could be affected in terms of git bisect, but I don't see too many risks.

jschwinger233 committed:
route: Clean up ip rules with "proto unspec"

This commit adds "removeStaleProxyRulesIPvX()" which removes any ip
rules with "proto unspec" to ensure upgrade/downgrade goes smoothly.

Scenario 1: upgrade from v1.15-old to v1.15-tip
  v1.15-old cilium installs ip rules with "proto unspec", then
  v1.15-tip will install "duplicate" ip rules with "proto kernel".
  This is the moment when "removeStaleProxyRulesIPvX()" plays a role,
  it cleans those "proto unspec" stale rules without breaking
  connectivity.

Scenario 2: downgrade from v1.15-tip to v1.15-old
  v1.15-tip has rules with "proto kernel". When v1.15-old tries to
  "ReplaceRule()" with "proto unspec", thanks to
  [this](https://github.com/cilium/cilium/blob/v1.15.3/pkg/datapath/linux/route/route_linux.go#L402),
  "ReplaceRule()" won't replace the rules because they already exist
  (with a different proto). This ensures connectivity can survive
  the downgrade too.

Scenario 3: upgrade from v1.15-tip to v1.16
  Since v1.15-tip installs correct rules with "proto kernel", v1.16
  will do nothing after confirming existance by "lookupRule()". It
  should be painless as well.

This is a v1.15-only commit because:
1. v1.14 is still using bpf/init.sh which sets rules with "proto kernel"
   properly;
2. v1.16 has been fixed to set "proto kernel";
3. v1.15-tip -> v1.16 upgrade has been discussed above without any
   issue;

Also please note that we don't have to clean up leftover ip routes with
"proto unspec", because we replace them via "route.Upsert()" which
replaces the old ones unconditionally, leaving no stale routes.

Signed-off-by: Zhichuan Liang <gray.liang@isovalent.com>

rgo3 commented on pkg/proxy/routes.go:
```diff
@@ -196,3 +207,26 @@ func removeFromProxyRoutesIPv6() error {
+		return fmt.Errorf("listing routing rules: %w", err)
+	}
+	for _, r := range rules {
+		if r.Protocol == unix.RTPROT_UNSPEC {
```
Should we also ensure that we only remove rules from the tables that concern us?

jschwinger233 replied:
Sure! Now it also checks `r.Table`:
```diff
- if r.Protocol == unix.RTPROT_UNSPEC {
+ if r.Protocol == unix.RTPROT_UNSPEC && slices.Contains(ciliumRouteTables, r.Table) {
```
